File: packages/agent-python/shieldrasp/detection/cmd.py
import re
from shieldrasp.taint.context import is_tainted, get_taint_context

SHELL_METACHARS = re.compile(r'[;|&`$><\n\\]|\$\(')

def detect_cmd_injection(command_args, ctx=None):
    if not ctx:
        ctx = get_taint_context()
        
    for arg in command_args:
        if isinstance(arg, str):
            matched = SHELL_METACHARS.search(arg) is not None
            tainted = is_tainted(arg)
            
            if matched and tainted:
                return {
                    "blocked": True,
                    "matched": True,
                    "type": "Command Injection",
                    "confidence": 0.99,
                    "cwe": "CWE-78",
                    "payload": arg
                }
    return {"blocked": False, "matched": False}

File: packages/agent/src/detection/cmd.ts
import { TaintContext, getTaintContext } from '../taint/context';

const SHELL_METACHARS = /[;|`$><\n\\&]|\$\(/;

export function detectCmdInjection(command: string, ctx?: TaintContext) {
  const matched = SHELL_METACHARS.test(command);
  const currentCtx = ctx || getTaintContext();
  if (!currentCtx) return { blocked: false, matched: false };

  let isTainted = false;
  let taintedValueMatched = '';

  for (const [taintedValue] of currentCtx.taintedObjects) {
    if (typeof taintedValue === 'string' && command.includes(taintedValue)) {
      isTainted = true;
      taintedValueMatched = taintedValue;
      break;
    }
  }

  if (isTainted) {
    // If it contains metacharacters, it's definitely a command injection attempt.
    if (matched || SHELL_METACHARS.test(taintedValueMatched)) {
      return {
        blocked: true, matched: true, attack_type: 'Command Injection', confidence: 0.99,
        cwe: 'CWE-78', payload: command, timestamp: Date.now()
      };
    }

    // Even without metachars, user input in a shell command is extremely risky.
    return {
      blocked: false, matched: true, attack_type: 'Tainted Command Execution', confidence: 0.60,
      cwe: 'CWE-78', payload: command, timestamp: Date.now(),
      details: 'User-controlled data was passed to a shell execution function.'
    };
  }

  return { blocked: false, matched: false };
}

